lib/models/addon.js

```diff
@@ -371,16 +391,76 @@ let addonProto = {
   },
 
   /**
-    Discovers all child addons of this addon and stores their names and
-    package.json contents in this.addonPackages as key-value pairs
-
-    @private
-    @method discoverAddons
+   * Discovers all child addons of this addon and an AddonInfo about
+   * each addon in this.addonPackages (keyed on addon name).
+   *
+   * Child addons include those from 'dependencies' (not devDependencies)
+   * and in-repo addons
+   *
+   * Any packageInfos that we find that are marked as not valid are excluded.
+   *
+   * @private
+   * @method discoverAddons
    */
   discoverAddons() {
-    let addonsList = this.addonDiscovery.discoverChildAddons(this);
+    let pkgInfo = this._packageInfoCache.getEntry(this.root);
+
+    let addonPackageList = [];
+
+    // There are cases where an addon can be created in memory and not actually
+    // have a root entry (or have one that is not actually pointing to a directory,
+    // like 'foo' in some of the tests. We don't want to crash, but want to let
+    // things continue even though they're empty.
+    if (pkgInfo) {
+
+      // The 'dependencyPackages' object contains ALL dependencies, not just those
+      // that are addons. We want only the addons.
+      if (pkgInfo.dependencyPackages) {
+        Object.keys(pkgInfo.dependencyPackages).forEach(name => {
+          if (name !== 'ember-cli') {
+            let dependencyInfo = pkgInfo.dependencyPackages[name];
+            if (dependencyInfo.isAddon()) {
+              addonPackageList.push(dependencyInfo);
+            }
+          }
+        });
+      }
 
-    this.addonPackages = this.addonDiscovery.addonPackages(addonsList);
+      if (pkgInfo.inRepoAddons) {
+        pkgInfo.inRepoAddons.forEach(addon => addonPackageList.push(addon));
```

Comment:
maybe we could just `addonPackageList.concat(pkgInfo.inRepoAddons)`?

Reply:
Definitely! Good catch. Done.